src/idpf.rs

```diff
@@ -0,0 +1,1852 @@
+                // Evaluate the IDPF starting from the cached data at a previously-computed
+                // node.
+                return eval_inner::<P, L, OUT_LEN>(
+                    bits,
+                    which_aggregator,
+                    public_share,
+                    cache_key.len(),
+                    seed,
+                    Choice::from(control_bit),
+                    prefix,
+                    cache,
+                );
+            }
+            cache_key = &cache_key[..cache_key.len() - 1];
+        }
+    }
+    // Evaluate starting from the root node.
+    eval_inner::<P, L, OUT_LEN>(
+        bits,
+        which_aggregator,
+        public_share,
+        0,
+        key.0,
+        Choice::from(which_aggregator as u8),
+        prefix,
+        cache,
+    )
+}
+
+/// IDPF public share used by Poplar1. This contains the lists of correction words used by all
+/// parties when evaluating the IDPF.
+#[derive(Debug, Clone)]
+pub struct IdpfPoplarPublicShare<FI, FL, const L: usize, const OUT_LEN: usize> {
+    /// Correction words for each inner node level.
+    inner_correction_words: Vec<IdpfPoplarCorrectionWord<FI, L, OUT_LEN>>,
+    /// Correction word for the leaf node level.
+    leaf_correction_word: IdpfPoplarCorrectionWord<FL, L, OUT_LEN>,
+}
+
+impl<FI, FL, const L: usize, const OUT_LEN: usize> ConstantTimeEq
+    for IdpfPoplarPublicShare<FI, FL, L, OUT_LEN>
+where
+    FI: ConstantTimeEq,
+    FL: ConstantTimeEq,
+{
+    fn ct_eq(&self, other: &Self) -> Choice {
+        self.inner_correction_words
+            .ct_eq(&other.inner_correction_words)
+            & self.leaf_correction_word.ct_eq(&other.leaf_correction_word)
+    }
+}
+
+impl<FI, FL, const L: usize, const OUT_LEN: usize> PartialEq
+    for IdpfPoplarPublicShare<FI, FL, L, OUT_LEN>
+where
+    FI: ConstantTimeEq,
+    FL: ConstantTimeEq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.ct_eq(other).into()
+    }
+}
+
+impl<FI, FL, const L: usize, const OUT_LEN: usize> Eq for IdpfPoplarPublicShare<FI, FL, L, OUT_LEN>
+where
+    FI: ConstantTimeEq,
+    FL: ConstantTimeEq,
+{
+}
+
+impl<FI, FL, const L: usize, const OUT_LEN: usize> Encode
+    for IdpfPoplarPublicShare<FI, FL, L, OUT_LEN>
+where
+    FI: Encode,
+    FL: Encode,
+{
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        let mut control_bits: BitVec<u8, Lsb0> =
+            BitVec::with_capacity(self.inner_correction_words.len() * 2 + 2);
+        for cws in self.inner_correction_words.iter() {
+            control_bits.extend(cws.control_bits.iter().map(|x| bool::from(*x)));
+        }
+        control_bits.extend(
+            self.leaf_correction_word
+                .control_bits
+                .iter()
+                .map(|x| bool::from(*x)),
+        );
+        control_bits.set_uninitialized(false);
+        let mut packed_control = control_bits.into_vec();
+        packed_control.reverse();
```

Comment:
If I reversed the semantic order of the bit indexes and switched from `Lsb0` to `Msb0`, I think that would cancel out and give us the combination of bit endianness and byte endianness we want.

Reply:
Wait, that wouldn't work for cases where we have unused bits, because the leading byte/most significant byte would have the wrong bits filled/empty. I think I'll have to ditch using `bitvec` here to avoid the `reverse()` call, and write the shifting and masking myself.

Reply:
(Sorry to do this.) Thanks for figuring out an alternative. The code is far more complex and therefore less clearly correct. In this instance, I would prefer (possibly) slower code that is more efficient. Would you mind reverting to using `reverse()`?